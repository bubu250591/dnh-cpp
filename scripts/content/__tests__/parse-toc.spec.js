import processMd, { getPostId } from "../parse-toc"
import fs from "fs"
import { resolve } from "path"

describe("parse toc", () => {
  it("parse correctly", () => {
    const file = fs.readFileSync(resolve(__dirname, "./data.txt"), "utf8")
    expect(processMd(file)).toMatchSnapshot()
  })

  it("simplify url", () => {
    // eslint-disable-next-line
    const from = "http://daynhauhoc.com/t/gioi-thieu-ve-series-tutorial-lap-trinh-c-danh-cho-nguoi-moi-bat-dau/29402"
    const to = 29402

    expect(getPostId(from)).toBe(to)
  })
})
