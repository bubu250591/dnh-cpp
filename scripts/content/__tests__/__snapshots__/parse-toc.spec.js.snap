exports[`parse toc parse correctly 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "id": 29402,
          "name": "0.0 Giới thiệu về series tutorial"
        },
        Object {
          "id": 29425,
          "name": "0.1 Giới thiệu về ngôn ngữ lập trình"
        },
        Object {
          "id": 29448,
          "name": "0.2 Giới thiệu về ngôn ngữ lập trình C++"
        },
        Object {
          "id": 29449,
          "name": "0.3 Các công đoạn phát triển của một chương trình C++"
        },
        Object {
          "id": 29503,
          "name": "0.4 Giới thiệu môi trường phát triển ứng dụng C++"
        },
        Object {
          "children": Array [
            Object {
              "id": 24768,
              "name": "Cài đặt Visual Studio 2015 để lập trình C++ trên Windows"
            },
            Object {
              "id": 24606,
              "name": "Hướng dẫn cài đặt Xcode để lập trình C++ trên Mac OS X"
            },
            Object {
              "id": 24563,
              "name": "Cài đặt Code::blocks để lập trình C/C++ trên Ubuntu/Linux"
            }
          ],
          "level": 2,
          "name": "0.5 Hướng dẫn cài đặt một số IDE để lập trình C++ trên các nền tảng Window,Mac Os, Linux"
        }
      ],
      "level": 1,
      "name": "Giới thiệu tổng quan khóa học"
    },
    Object {
      "children": Array [
        Object {
          "id": 29504,
          "name": "1.0 Viết chương trình đầu tiên"
        },
        Object {
          "id": 29527,
          "name": "1.1 Cấu trúc cơ bản của một chương trình C++"
        },
        Object {
          "id": 29575,
          "name": "1.2 Lệnh, khối lệnh, từ khóa"
        },
        Object {
          "id": 29796,
          "name": "1.3 Sử dụng các lệnh liên quan đến xuất dữ liệu"
        },
        Object {
          "id": 29935,
          "name": "1.4 Biến và các kiểu dữ liệu trong C++"
        },
        Object {
          "id": 29936,
          "name": "1.5 Nhập và xuất dữ liệu"
        },
        Object {
          "id": 29938,
          "name": "1.6 Hằng số"
        },
        Object {
          "id": 29939,
          "name": "1.7 Phạm vi của biến"
        },
        Object {
          "id": 29944,
          "name": "1.8 Các phép toán cơ bản"
        },
        Object {
          "id": 29945,
          "name": "1.9 Toán tử tăng giảm"
        },
        Object {
          "id": 29574,
          "name": "1.10 Độ ưu tiên của các toán tử"
        }
      ],
      "level": 1,
      "name": "C++ cơ bản"
    },
    Object {
      "children": Array [
        Object {
          "id": 29992,
          "name": "2.0 Boolean"
        },
        Object {
          "id": 30007,
          "name": "2.1 Giới thiệu một số cấu trúc điều khiển"
        },
        Object {
          "id": 30008,
          "name": "2.2 if statements"
        },
        Object {
          "id": 30009,
          "name": "2.3 switch case statements"
        }
      ],
      "level": 1,
      "name": "Cấu trúc rẽ nhánh"
    },
    Object {
      "children": Array [
        Object {
          "id": 30011,
          "name": "3.0 Vòng lặp while"
        },
        Object {
          "id": 30012,
          "name": "3.1 Vòng lặp do-while"
        },
        Object {
          "id": 30013,
          "name": "3.2 Vòng lặp for"
        },
        Object {
          "id": 30014,
          "name": "3.3 từ khóa break và continue"
        }
      ],
      "level": 1,
      "name": "Cấu trúc vòng lặp"
    },
    Object {
      "children": Array [
        Object {
          "id": 30016,
          "name": "4.0 Sử dụng thư viện cstdint"
        },
        Object {
          "id": 30018,
          "name": "4.1 Kiểu kí tự"
        },
        Object {
          "id": 30019,
          "name": "4.2 Ép kiểu dữ liệu"
        },
        Object {
          "id": 30020,
          "name": "4.3 auto và decltype"
        },
        Object {
          "id": 30022,
          "name": "4.4 Địa chỉ của biến"
        }
      ],
      "level": 1,
      "name": "Nâng cao về biến, kiểu dữ liệu"
    },
    Object {
      "children": Array [
        Object {
          "id": 30080,
          "name": "5.0 Mảng một chiều"
        },
        Object {
          "id": 30082,
          "name": "5.1 Các thao tác cơ bản với mảng một chiều"
        },
        Object {
          "id": 30083,
          "name": "5.2 Thư viện array trong STL"
        },
        Object {
          "id": 30085,
          "name": "5.3 Mảng hai chiều"
        },
        Object {
          "id": 30086,
          "name": "5.4 Các thao tác cơ bản với mảng hai chiều"
        }
      ],
      "level": 1,
      "name": "Kiểu dữ liệu mảng"
    },
    Object {
      "children": Array [
        Object {
          "id": 30090,
          "name": "6.0 Mảng kí tự"
        },
        Object {
          "id": 30091,
          "name": "6.1 Các thao tác cơ bản với mảng kí tự"
        },
        Object {
          "id": 30092,
          "name": "6.2 Thư viện string trong C++"
        }
      ],
      "level": 1,
      "name": "Kiểu chuỗi kí tự"
    },
    Object {
      "children": Array [
        Object {
          "id": 30132,
          "name": "7.0 Làm quen với khái niệm function"
        },
        Object {
          "id": 30133,
          "name": "7.1 Thiết kế chương trình dựa trên khái niệm function"
        },
        Object {
          "id": 30319,
          "name": "7.2 Truyền đối số cho hàm là giá trị hoặc tham chiếu"
        },
        Object {
          "id": 30381,
          "name": "7.3 Hàm trả về giá trị hoặc tham chiếu"
        },
        Object {
          "id": 30382,
          "name": "7.4 Tham số mặc định, tham số hằng"
        },
        Object {
          "id": 30477,
          "name": "7.5 Function overloading"
        }
      ],
      "level": 1,
      "name": "Cơ bản về Function"
    },
    Object {
      "children": Array [
        Object {
          "id": 30541,
          "name": "8.0 Con trỏ"
        },
        Object {
          "id": 30699,
          "name": "8.1 Toán tử tăng, giảm dùng cho con trỏ"
        },
        Object {
          "id": 30788,
          "name": "8.2 Con trỏ và mảng một chiều"
        },
        Object {
          "id": 30876,
          "name": "8.3 Con trỏ và mảng kí tự"
        },
        Object {
          "id": 31145,
          "name": "8.4 Cấp phát bộ nhớ động"
        },
        Object {
          "id": 31355,
          "name": "8.5 Con trỏ và hằng"
        },
        Object {
          "id": 31625,
          "name": "8.6 Con trỏ void"
        },
        Object {
          "id": 32159,
          "name": "8.7 Con trỏ trỏ đến con trỏ"
        },
        Object {
          "id": 31810,
          "name": "8.8 Con trỏ và hàm"
        },
        Object {
          "id": 31959,
          "name": "8.9 Con trỏ hàm"
        },
        Object {
          "id": 32101,
          "name": "8.10 Phân loại các vùng nhớ (stack & heap ...)"
        },
        Object {
          "id": 32133,
          "name": "8.11 Các vấn đề thường gặp khi sử dụng con trỏ"
        }
      ],
      "level": 1,
      "name": "Con trỏ"
    },
    Object {
      "children": Array [
        Object {
          "id": 32760,
          "name": "9.0 Kiểu liệt kê (enum)"
        },
        Object {
          "id": 33291,
          "name": "9.1 Structs"
        },
        Object {
          "level": 2,
          "name": "9.2 Struct and pointer"
        }
      ],
      "level": 1,
      "name": "Kiểu dữ liệu tự định nghĩa"
    },
    Object {
      "children": Array [
        Object {
          "level": 2,
          "name": "10.0 File, Stream trong C++"
        },
        Object {
          "level": 2,
          "name": "10.1 Đọc dữ liệu từ file trong C++"
        },
        Object {
          "level": 2,
          "name": "10.2 Ghi dữ liệu vào file trong C++"
        },
        Object {
          "level": 2,
          "name": "10.3 Stream for string C++"
        },
        Object {
          "level": 2,
          "name": "10.4 Stream states & Stream validation"
        }
      ],
      "level": 1,
      "name": "Nhập, xuất (Input & Output)"
    },
    Object {
      "children": Array [
        Object {
          "level": 2,
          "name": "11.0 Vector"
        },
        Object {
          "level": 2,
          "name": "11.1 Stack (Ngăn xếp)"
        },
        Object {
          "level": 2,
          "name": "11.2 Queue (Hàng đợi)"
        }
      ],
      "level": 1,
      "name": "Một số thư viện thường dùng trong STL"
    },
    Object {
      "children": Array [
        Object {
          "level": 2,
          "name": "12.0 auto_ptr"
        },
        Object {
          "level": 2,
          "name": "12.1 shared_ptr"
        },
        Object {
          "level": 2,
          "name": "12.2 unique_ptr"
        }
      ],
      "level": 1,
      "name": "Auto pointer"
    },
    Object {
      "children": Array [
        Object {
          "level": 2,
          "name": "13.0 Biến static"
        },
        Object {
          "level": 2,
          "name": "13.1 Biến toàn cục, tại sao không nên sử dụng nhiều biến toàn cục"
        },
        Object {
          "level": 2,
          "name": "13.2 Viết chương trình với nhiều file"
        },
        Object {
          "level": 2,
          "name": "13.3 Tạo và sử dụng thư viện liên kết tĩnh"
        },
        Object {
          "level": 2,
          "name": "13.4 Tạo và sử dụng thư viện liên kết động"
        }
      ],
      "level": 1,
      "name": "Quản lý mã nguồn"
    },
    Object {
      "children": Array [
        Object {
          "level": 2,
          "name": "14.0 Function template"
        },
        Object {
          "level": 2,
          "name": "14.1 Lambda expression"
        },
        Object {
          "level": 2,
          "name": "14.2 Xử lý ngoại lệ"
        }
      ],
      "level": 1,
      "name": "Một số feature trong C++11, C++14"
    }
  ],
  "level": 0,
  "name": "Khóa học C++",
  "toggled": true
}
`;
